src/client-side-encryption/clientEncryption.ts

```diff
@@ -0,0 +1,1015 @@
+   * }
+   * ```
+   *
+   * @example
+   * ```ts
+   * // Decrypting value with async/await API
+   * async function decryptMyValue(value) {
+   *   return clientEncryption.decrypt(value);
+   * }
+   * ```
+   */
+  decrypt(value: Binary, callback?: Callback<Document>): Promise<Document> | void {
+    const valueBuffer = serialize({ v: value });
+    const context = this._mongoCrypt.makeExplicitDecryptionContext(valueBuffer);
+
+    const stateMachine = new StateMachine({
+      proxyOptions: this._proxyOptions,
+      tlsOptions: this._tlsOptions
+    });
+
+    // @ts-expect-error We did not convert promiseOrCallback to TS
+    return promiseOrCallback(callback, cb => {
+      stateMachine.execute(this, context, (err, result) => {
+        if (err || !result) {
+          cb(err, null);
+          return;
+        }
+
+        cb(null, result.v);
+      });
+    });
+  }
+
+  /**
+   * Ask the user for KMS credentials.
+   *
+   * This returns anything that looks like the kmsProviders original input
+   * option. It can be empty, and any provider specified here will override
+   * the original ones.
+   */
+  async askForKMSCredentials(): Promise<KMSProviders> {
+    return loadCredentials(this._kmsProviders);
+  }
+
+  static get libmongocryptVersion() {
+    return ClientEncryption.getMongoCrypt().libmongocryptVersion;
+  }
+
+  /**
+   * A helper that perform explicit encryption of values and expressions.
+   * Explicitly encrypt a provided value. Note that either `options.keyId` or `options.keyAltName` must
+   * be specified. Specifying both `options.keyId` and `options.keyAltName` is considered an error.
+   *
+   * @param value - The value that you wish to encrypt. Must be of a type that can be serialized into BSON
+   * @param expressionMode - a boolean that indicates whether or not to encrypt the value as an expression
+   * @param options - options to pass to encrypt
+   * @returns the raw result of the call to stateMachine.execute().  When expressionMode is set to true, the return
+   *          value will be a bson document.  When false, the value will be a BSON Binary.
+   *
+   */
+  private async _encrypt(
+    value: unknown,
+    expressionMode: boolean,
+    options: ClientEncryptionEncryptOptions
+  ): Promise<Binary> {
+    const { algorithm, keyId, keyAltName, contentionFactor, queryType, rangeOptions } = options;
+    const contextOptions: ExplicitEncryptionContextOptions = {
+      expressionMode,
+      algorithm
+    };
+    if (keyId) {
+      contextOptions.keyId = keyId.buffer;
+    }
+    if (keyAltName) {
+      if (keyId) {
+        throw new TypeError(`"options" cannot contain both "keyId" and "keyAltName"`);
+      }
+      if (typeof keyAltName !== 'string') {
+        throw new TypeError(
+          `"options.keyAltName" must be of type string, but was of type ${typeof keyAltName}`
+        );
+      }
+
+      contextOptions.keyAltName = serialize({ keyAltName });
+    }
```

Comment:
I refactored options construction here because the old approach gives TS errors.  For example, keyId.

```typescript
const contextOptions: ExplicitEncryptionContextOptions = Object.assign({}, options);
if (options.keyId) {
  contextOptions.keyId = options.keyId.buffer;
}
```

In the above, TS complains in the first line because we assign a `Binary` to the `keyId` option, which accepts a buffer, only to re-assign it in the if statement.  Most of the options were impacted by this issue.